Added overview page for new "Event Messaging System" section (#2210)
o3de/o3de.org: content/docs/user-guide/programming/messaging/overview.md

---
linktitle: Overview
title: Overview of Event Messaging Systems
description: Use event buses (EBus), AZ::Interface, and AZ::Event to dispatch messages between systems in Open 3D Engine (O3DE). 
weight: 100
---

> Event messaging systems are essential to a game engine because they allow each system to communicate with each other. **Open 3D Engine (O3DE)** uses a few types of communication methods, which are separated into the following modules: 

Reviewer: ```suggestion
Event messaging systems are essential to a game engine because they allow systems to communicate with each other. **Open 3D Engine (O3DE)** uses a few types of communication methods, which are separated into the following modules: 
```

- **Event Bus (EBus)**: A single global bus that all modules can use to invoke requests and dispatch messages, respectively known as *request* and *notification* buses.

> - `AZ::Interface`: A C++ template class for a global interface that other components can invoke requests from. The equivalent of the EBus sytem's request bus.

Reviewer: ```suggestion
- `AZ::Interface`: A C++ template class for a global interface that other components can invoke requests from, equivalent to the EBus sytem's request bus.
```
Suggestion to avoid a sentence fragment.

> - `AZ::Event`: A C++ template class that can publish single-value messages, which other components can subscribe to. The equivalent of the EBus system's notification bus.

Reviewer: ```suggestion
- `AZ::Event`: A C++ template class that can publish single-value messages, which other components can subscribe to, equivalent to the EBus system's notification bus.
```

> Some of the functionality of these three modules overlap: `AZ::Interface` provides a request interface, `AZ::Event` provides a publish-subscribe interface, and EBus provides both. However, each module has strengths and weaknesses that make them more suitable for different scenarios.

Reviewer 2: **Alternatively**, you can write around these issues. Consider the following. IMO, this makes the first sentence a bit cleaner and easier to follow the subject-verb agreement. And in the second sentence, pluralizing the subject ("each module" > "all three modules") avoids any confusion resulting from mixing a singular subject with other plural nouns ("strengths and weaknesses").

```suggestion
The functionality of these three modules overlaps in some ways: `AZ::Interface` provides a request interface, `AZ::Event` provides a publish-subscribe interface, and EBus provides both. However, all three modules have strengths and weaknesses that make them more suitable for different scenarios.
```

> Some of the functionality of these three modules overlap: `AZ::Interface` provides a request interface, `AZ::Event` provides a publish-subscribe interface, and EBus provides both. However, each module has strengths and weaknesses that make them more suitable for different scenarios.

Author: @paucom Thanks for further elaborating on the proper grammar here! I'll accept your alternative suggestion, i agree the wording is less confusing.

> Some of the functionality of these three modules overlaps: `AZ::Interface` provides a request interface, `AZ::Event` provides a publish-subscribe interface, and EBus provides both. However, each module has strengths and weaknesses that make it more suitable for different scenarios.

Reviewer: ```suggestion
Some of the functionality of these three modules overlap: `AZ::Interface` provides a request interface, `AZ::Event` provides a publish-subscribe interface, and EBus provides both. However, each module has strengths and weaknesses that make them more suitable for different scenarios.
```

The following table compares some key characteristics of EBus, `AZ::Interface`, and `AZ::Event`. You can find all three event messaging modules throughout O3DE's code base.

| | EBus | `AZ::Interface` | `AZ::Event` |
| --- | --- | --- | --- |

> | **Communication model** |  EBus is a general purpose event messaging system whose design focuses on abstraction and decoupling systems. It allows many source components to provide requests or publish messages. It also allows many listener components to invoke those requests or subscribe to those messages---all without the components having to know about each other. Unlike `AZ::Event`, a listener can subscribe to an EBus specialization even if the EBus doesn't exist yet. | An `AZ::Interface<T>` template specialization defines a global singleton-like request interface that other systems can invoke. To access the request interface, other systems can call `AZ::Interface<T>.Get()`. | `AZ::Event` events are defined as members of a component. As such, other components must have a reference to the component to subscribe to the event. To listen and process an event, subscriber components must implement a handler. A single handler can connect to only one event, but an event can have multiple handlers.<br><br>Some patterns for cross-entity communication include:<ul><li>Exposing the `AZ::Event` to an `AZ::Interface`.</li><li>Calling `SceneQuery()` and `AZ::IVisibility` to get a reference to the entity.</li></ul> |

Reviewer: ```suggestion
| **Communication model** |  EBus is a general purpose event messaging system whose design focuses on abstraction and decoupling systems. It allows many source components to provide requests or publish messages. It also allows many listener components to invoke those requests or subscribe to those messages---all without the components needing to know about each other. Unlike `AZ::Event`, a listener can subscribe to an EBus specialization even if the EBus doesn't exist yet. | An `AZ::Interface<T>` template specialization defines a global singleton-like request interface that other systems can invoke. To access the request interface, other systems can call `AZ::Interface<T>.Get()`. | `AZ::Event` events are defined as members of a component. As such, other components must have a reference to the component to subscribe to the event. To listen and process an event, subscriber components must implement a handler. A single handler can connect to only one event, but an event can have multiple handlers.<br><br>Some patterns for cross-entity communication include:<ul><li>Exposing the `AZ::Event` to an `AZ::Interface`.</li><li>Calling `SceneQuery()` and `AZ::IVisibility` to get a reference to the entity.</li></ul> |
```

Reviewer: "Having to" -> "needing to" was the only suggestion here.

> | Communication model | An `AZ::Interface<T>` template specialization defines a global singleton-like request interface, which other systems can invoke. Other systems can simply access the request interface by calling `AZ::Interface<T>.Get()`.  | An `AZ::Event` instance must be attached to a source component. Therefore, other components must have a reference to the source component to subscribe to the event. Other components subscribe to an event by implementing a handler that will process the event when it's signaled. An event can be signaled by calling `Signal()`, which subscribers can listen to and process via their event handler. A handler can only connect to one event, but an event can have multiple handlers. Events are useful for components talking to other components within the same entity. For cross-entity communication, you can implement `AZ::Event`s on an `AZ::Interface`, or use an EBus. | EBus is a general purpose event messaging system whose design focuses on abstraction and decoupling systems. It allows many source components to provide requests or publish messages, and many listener components to invoke those requests or subscribe to those messages - all without the components having to know about each other. Unlike `AZ::Event`, a listener can subscribe to an EBus specialization even if it doesn't exist yet. |

Reviewer 3: 'For cross-entity communication, you can implement `AZ::Event`s on an `AZ::Interface`'
You just need some way to access the event. Scene queries and the AZ::IVisibility both allow efficient access to entities within an area of interest. The AZ::Events don't need to be limited to those that exist directly on an AZ::Interface.

> | **Lifetime** | There can only be a single instance of an `AZ::Interface<T>` specialization at a time. It initializes when the application launches and gets destroyed when the application is terminated. | There can be many instances of an `AZ::Event<T>` specialization. Each instance must be attached to a component. It initializes when the component is created and gets destroyed when the component is destroyed. | EBus is a global singleton. It initializes when the application launches and gets destroyed when the application terminates. |

Author: Yes, "an instance" can replace "it". And the ambiguous "it" refers to EBus. I changed the sentence to read as: 
> EBus is a global singleton that initializes when...

> | **Lifetime** | EBus is a global singleton that initializes when the application launches and gets destroyed when the application terminates. | There can be only a single instance of an `AZ::Interface<T>` specialization at a time. An instance initializes when the application launches and gets destroyed when the application is terminated. | There can be many instances of an `AZ::Event<T>` specialization. Each instance must be attached to a component. An instance initializes when the component is created and gets destroyed when the component is destroyed. |

Reviewer: ```suggestion
| **Lifetime** | EBus is a global singleton that initializes when the application launches and gets destroyed when the application terminates. | There can be only a single instance of an `AZ::Interface<T>` specialization at a time. An instance initializes when the application launches and gets destroyed when the application terminates. | There can be many instances of an `AZ::Event<T>` specialization. Each instance must be attached to a component. An instance initializes when the component is created and gets destroyed when the component is destroyed. |
```
Change to match the usage in the previous sentence.

> | Scripting |  |  | Unlike `AZ::Interface`, only EBus can connect to scripting. This allows your component to expose its functionality to O3DE's scripting modules, such as Script Canvas and Lua. |

Reviewer 3: There exists script-bind support for AZ::Event. It may be good to include that information here.

> | **Advantages** | EBus is O3DE's most flexible and powerful event messaging system. Any system can connect to it and communicate with any other system by simply providing a system's EBus name. EBus is generally used for situations where event flow is more important than the source of the event. | Compared to EBus, `AZ::Interface` has improved runtime performance, improved debuggability, and compatibility with code autocomplete. | Compared to EBus, `AZ::Event` has improved runtime performance, allows simpler syntax to implement, uses fewer files, and removes aggregate interfaces where a handler only cares about a subset of events. |

Reviewer: ```suggestion
| **Advantages** | EBus is O3DE's most flexible and powerful event messaging system. Any system can connect to it and communicate with any other system by simply providing a system's EBus name. EBus is generally used for situations where event flow is more important than the source of the event. | Compared to EBus, `AZ::Interface` has improved runtime performance, improved debuggability, and compatibility with code autocompletion. | Compared to EBus, `AZ::Event` has improved runtime performance, uses simpler implementation syntax, has fewer files, and removes aggregate interfaces whereas a handler only manages a subset of events. |
```

> | **Examples** | A spawner-spawnee system. The spawner interface contains functions to manage spawnees. If we want another class to spawn something, that class can make a request via  `AZ::Interface<ISpawner>`. By calling `AZ::Interface<ISpawner>.Get()`, other components can invoke requests from the spawner. | In the networking layer, when a remote process call (RPC) is sent, it signals an `AZ::Event<NetworkEntityRpcMessage&>`. Then, a connected handler listens for that signal change and processes the `NetworkEntityRpcMessage`. | EBus examples are similar to `AZ::Interface` and `AZ::Event`. In most cases, using those two is more preferable due to their optimizations. However, you must use EBus to add script-binding support for request busses.  |

Reviewer 2: ```suggestion
| **Examples** | A spawner-spawnee system. The spawner interface contains functions to manage spawnees. If you want another class to spawn something, that class can make a request via  `AZ::Interface<ISpawner>`. By calling `AZ::Interface<ISpawner>.Get()`, other components can invoke requests from the spawner. | In the networking layer, when a remote process call (RPC) is sent, it signals an `AZ::Event<NetworkEntityRpcMessage&>`. Then, a connected handler listens for that signal change and processes the `NetworkEntityRpcMessage`. | EBus examples are similar to `AZ::Interface` and `AZ::Event`. In most cases, using those two is preferable due to their optimizations. However, you must use EBus to add script binding support for request buses. |
```
* Changed ["we" to "you"](https://www.o3de.org/docs/contributing/to-docs/style-guide/guidance/#voice-and-tone)
* Should "remote process call" be "remote procedure call"? See [Remote procedure call](https://en.wikipedia.org/wiki/Remote_procedure_call) (Wikipedia).

Author: Yes, it should be called "remote procedure call". Thanks for catching!

> | **Examples** | EBus examples are similar to `AZ::Interface` and `AZ::Event`. In most cases, using those two is more preferable due to their optimizations. However, you must use EBus to add script-binding support for request busses.  | A spawner-spawnee system. The spawner interface contains functions to manage spawnees. If we want another class to spawn something, that class can make a request via  `AZ::Interface<ISpawner>`. By calling `AZ::Interface<ISpawner>.Get()`, other components can invoke requests from the spawner. | In the networking layer, when a remote process call (RPC) is sent, it signals an `AZ::Event<NetworkEntityRpcMessage&>`. Then, a connected handler listens for that signal change and processes the `NetworkEntityRpcMessage`. |

Reviewer: ```suggestion
| **Examples** | EBus examples are similar to `AZ::Interface` and `AZ::Event`. In most cases, using `AZ::Interface` or `AZ::Event` is preferable due to  optimizations. However, you must use EBus to add script-binding support for request buses.  | A spawner-spawnee system. The spawner interface contains functions to manage spawnees. If we want another class to spawn something, that class can make a request via  `AZ::Interface<ISpawner>`. By calling `AZ::Interface<ISpawner>.Get()`, other components can invoke requests from the spawner. | In the networking layer, when a remote process call (RPC) is sent, it signals an `AZ::Event<NetworkEntityRpcMessage&>`. Then, a connected handler listens for that signal change and processes the `NetworkEntityRpcMessage`. |
```
